### PR TITLE
Created roles guard

### PR DESCRIPTION
**Pull request by Lunessr**

### CODE REVIEW

File: src/decorators/roles.decorator.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import { SetMetadata } from "@nestjs/common";
+import { Role } from "../enums/role";
+
+export const ROLES_KEY = "roles";
```
Comment (rudenkose) [inline src/decorators/roles.decorator.ts:4]: this have to be moved to constants

File: src/guards/roles.guard.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  canActivate(context: ExecutionContext): boolean {
+    const requiredRoles = this.reflector.getAllAndOverride<Role[]>(ROLES_KEY, [
+      context.getHandler(),
+      context.getClass(),
+    ]);
+    if (!requiredRoles) {
+      return true;
+    }
+    const { user } = context.switchToHttp().getRequest();
+    return requiredRoles.some((role) => user.role?.includes(role));
```
Comment (rudenkose) [inline src/guards/roles.guard.ts:19]: why do we need `user.role?` ? in the end?
do we have the case when user have no roles?

File: src/users/customer.user.service.ts [hunk 3]
```diff
@@ -28,16 +28,17 @@ export class CustomerUsersService implements IUserService {
...
   async findAll(parameters: UserParameters): Promise<User[]> {
-    const {
+    let {
       filter,
       sortBy,
       direction = "DESC",
       limit = 3,
       skip = 0,
     } = parameters;
+    filter.role = "Customer";
```
Comment (rudenkose) [inline src/users/customer.user.service.ts:38]: You have to use `enum`

File: src/users/users.repository.ts [hunk 4]
```diff
@@ -17,31 +17,38 @@ export class UsersRepo {
   }
 
   public async findAll(parameters: UserParameters): Promise<User[]> {
-    let searchingParameters;
+    let where = {};
+    let order = {};
+    let take;
+    let skip;
```
Comment (rudenkose) [inline src/users/users.repository.ts:23]: why can't we use empty object as initial state for our _find request params?_

**Review by rudenkose — CHANGES_REQUESTED** [review]
Looks good, but a few comments inside

File: src/users/admin.users.service.ts [hunk 5]
```diff
@@ -22,9 +22,9 @@ export class AdminUsersService implements IUserService {
   }
 
   async findAll(parameters: UserParameters): Promise<User[]> {
-    const {
+    let {
```
Comment (rudenkose) [inline src/users/admin.users.service.ts:25]: why do we need to change it to let?

File: src/users/admin.users.service.ts [hunk 6]
```diff
@@ -81,8 +81,12 @@ export class AdminUsersService implements IUserService {
     if (userWithEmail === null || !userPassword) {
       throw new Error(ERROR_MESSAGES.INCORRECT_LOGIN);
     }
-    return jwt.sign({ id: userWithEmail.id }, process.env.SECRET_KEY, {
-      expiresIn: "1h",
-    });
+    return jwt.sign(
+      { id: userWithEmail.id, role: userWithEmail.role },
```
Comment (rudenkose) [inline src/users/admin.users.service.ts:85]: hah, nice catch!

File: src/users/customer.user.service.ts [hunk 7]
```diff
@@ -28,16 +28,17 @@ export class CustomerUsersService implements IUserService {
   }
 
   async findAll(parameters: UserParameters): Promise<User[]> {
-    const {
+    let {
```
Comment (rudenkose) [inline src/users/customer.user.service.ts:31]: why let again?

**Review by rudenkose — APPROVED** [review]
Looks good!
minor comments inside

File: src/users/admin.users.service.ts (hunk 5 above)
Comment (Lunessr) [reply src/users/admin.users.service.ts:25]: we don't need this "let". I forgot to rewrite it to "const"

**Pull request merged**